lib/ecto/adapters/sql.ex

```diff
@@ -641,33 +641,57 @@ defmodule Ecto.Adapters.SQL do
   def insert_all(adapter_meta, schema_meta, conn, header, rows, on_conflict, returning, opts) do
     %{source: source, prefix: prefix} = schema_meta
     {_, conflict_params, _} = on_conflict
-    {rows, params} = unzip_inserts(header, rows)
-    sql = conn.insert(prefix, source, header, rows, on_conflict, returning)
+    {rows, {params, placeholder_index_map, counter_offset}} = unzip_inserts(header, rows)
+
+    sql = conn.insert(prefix, source, header, rows, on_conflict, returning, counter_offset)
 
     opts = [{:cache_statement, "ecto_insert_all_#{source}"} | opts]
 
+    placeholder_map = Keyword.get(opts, :placeholders, %{})
+    placeholder_values = placeholder_index_map
+    |> Enum.map(fn {ph_key, ph_idx} ->
+      {ph_idx, Map.fetch!(placeholder_map, ph_key)}
+    end)
+    |> List.keysort(0)
+    |> Enum.map(&elem(&1, 1))
+
+    all_params = placeholder_values ++ Enum.reverse(params) ++ conflict_params
```

Comment:
Small optimization we didn't apply before:

```suggestion
    all_params = placeholder_values ++ Enum.reverse(params, conflict_params)
```